File: 0883-116-2467-most-profitable-path-in-a-tree/0883-116-2467-most-profitable-path-in-a-tree.py
from typing import List

class Solution:
    def __init__(self):
        self.BobT = []
        self.parent = []

    def BobDfs(self, n: int, p: int, adj: List[List[int]]):
        self.parent[n] = p
        for ch in adj[n]:
            if ch == p:
                continue
            self.BobDfs(ch, n, adj)

    def AliceDfs(self, n: int, p: int, adj: List[List[int]], AliceT: int, amount: List[int]) -> int:
        AliceM = 0
        if AliceT < self.BobT[n]:
            AliceM = amount[n]
        elif AliceT == self.BobT[n]:
            AliceM = amount[n] // 2

        if len(adj[n]) == 1 and n != 0:
            return AliceM

        mx = -1e9
        for ch in adj[n]:
            if ch == p:
                continue
            mx = max(mx, self.AliceDfs(ch, n, adj, AliceT + 1, amount))
        return mx + AliceM

    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        n = len(edges) + 1
        adj = [[] for _ in range(n)]
        for it in edges:
            adj[it[0]].append(it[1])
            adj[it[1]].append(it[0])

        self.BobT = [1e9] * n
        self.parent = [1e9] * n
        self.BobDfs(0, -1, adj)

        time = 0
        i = bob
        while i != -1:
            self.BobT[i] = time
            time += 1
            i = self.parent[i]

        return self.AliceDfs(0, -1, adj, 0, amount)

File: 0883-116-2467-most-profitable-path-in-a-tree/0883-116-2467-most-profitable-path-in-a-tree.java
import java.util.*;

class Solution {
    private int[] BobT;
    private int[] parent;

    private void BobDfs(int n, int p, List<List<Integer>> adj) {
        parent[n] = p;
        for (int ch : adj.get(n)) {
            if (ch == p) continue;
            BobDfs(ch, n, adj);
        }
    }

    private int AliceDfs(int n, int p, List<List<Integer>> adj, int AliceT, int[] amount) {
        int AliceM = 0;
        if (AliceT < BobT[n]) AliceM = amount[n];
        else if (AliceT == BobT[n]) AliceM = amount[n] / 2;

        if (adj.get(n).size() == 1 && n != 0) return AliceM;

        int mx = (int) -1e9;
        for (int ch : adj.get(n)) {
            if (ch == p) continue;
            mx = Math.max(mx, AliceDfs(ch, n, adj, AliceT + 1, amount));
        }
        return mx + AliceM;
    }

    public int mostProfitablePath(int[][] edges, int bob, int[] amount) {
        int n = edges.length + 1;
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] it : edges) {
            adj.get(it[0]).add(it[1]);
            adj.get(it[1]).add(it[0]);
        }

        BobT = new int[n];
        parent = new int[n];
        Arrays.fill(BobT, (int) 1e9);
        Arrays.fill(parent, (int) 1e9);
        BobDfs(0, -1, adj);

        int time = 0;
        for (int i = bob; i != -1; i = parent[i]) {
            BobT[i] = time++;
        }

        return AliceDfs(0, -1, adj, 0, amount);
    }
}